core/src/main/java/org/infinispan/protostream/impl/LazyByteArrayOutputStream.java

```diff
@@ -0,0 +1,83 @@
+package org.infinispan.protostream.impl;
+
+import java.io.OutputStream;
+
+import net.jcip.annotations.NotThreadSafe;
+
+@NotThreadSafe
+public class LazyByteArrayOutputStream extends OutputStream implements org.infinispan.protostream.LazyByteArrayOutputStream {
+
+   static final int DEFAULT_DOUBLING_SIZE = 4 * 1024 * 1024; // 4MB
+
+   public LazyByteArrayOutputStream() {
+      this(32);
+   }
+
+   public LazyByteArrayOutputStream(int capacity) {
+      if (capacity < 0)
+         throw new IllegalArgumentException("Negative initial capacity: " + capacity);
+      this.buf = new byte[capacity];
+   }
+
+   byte[] buf;
+   int pos = 0;
+
+   @Override
+   public void write(int b) {
+      ensureCapacity(pos + 1);
+      buf[pos] = (byte) b;
+      pos += 1;
```

Comment:
you're computing the new position twice in all the methods `write(byte)`and `write(byte[])`. I'm not sure if it will make any difference:
```java
      var newpos = pos + 1;
      ensureCapacity(newpos);
      buf[pos] = (byte) b;
      pos = newpos;
```